### ISSUE TITLE
Saving a document subset breaks defaults on further calls to Model.create

### ISSUE THREAD

**Issue by sebastian-nowak** [Automattic/mongoose#6892]
It appears to be a pretty serious issue, it took me a while to isolate it and come up with a repro script. It looks like calling `save()` on a document that was retrieved with `.select()` somehow affects further calls to `Model.create`.

```js
#!/usr/bin/env node
(async function () {
  try {
    const mongoose = require('mongoose');
    mongoose.connect('mongodb://localhost:27017/test', { useNewUrlParser: true });

    // We start by creating a model that has a field that defaults to null. We also
    // add a validator that will fail when that field is undefined, meaning that the
    // default value didn't kick in
    const schema = new mongoose.Schema({
      field: {
        type: String,
        default: null,
        required: [
          function () { return this && this.field === undefined; },
          'some err'
        ]
      }
    });
    const Model = mongoose.model('test', schema);

    // This part is fine, it's here just to demonstrate that it works. First model
    // instance will be created successfully, and if you inspect the call to `field`
    // validator you'll see that it received a {_id: something, field: null} object as
    // `this` so that default value was correctly applied. This is the expected behavior.
    console.log('Creating instance1');
    const instance1 = await Model.create({});
    console.log('Instance1 created');

    // This is the strange part. After doing this, the above snippet will no longer work.
    // Note the call to `select()` - if you remove it, the bug isn't triggered.
    console.log('Saving something');
    const instance1Retrieved = await Model.findById(instance1._id).select({_id: 1});
    await instance1Retrieved.save();
    console.log('Something saved');

    // So let's try creating a document again. This time it's going to fail,
    // ValidationError will be thrown. If you inspect the call to `field` validator
    // you'll notice that this time it received just a {_id: something} object,
    // defaults were ignored.
    console.log('Creating instance2');
    const instance2 = await Model.create({});
    console.log('Instance2 created');

    console.log('This message won\'t appear');
  } catch (err) {
    console.log(err);
    process.exit(0);
  }
})();
```

**Comment by sebastian-nowak** [comment]
Hey @vkarpov15,

I did some digging and it looks like you've introduced this bug in https://github.com/Automattic/mongoose/commit/751f4bb0d616054f1ec3b17f4fab48cac3116a57.

It's a bug in the validation code and it can be also reproduced by replacing `Model.create` with `new Model()` followed by a call to `.validate()`.

This part is what's directly causing the second validation to fail:

```js
// `$cachedRequired` gets set in `_evaluateRequiredFunctions()` so we
// don't call required functions multiple times in one validate call
// See gh-6801
if ('$cachedRequired' in _this) {
  const res = _this.$cachedRequired ? _this.checkRequired(v, this) : true;
  delete _this.$cachedRequired;
  return res;
} else if (typeof required === 'function') {
  return required.apply(this) ? _this.checkRequired(v, this) : true;
}
```

When validating the `instance2` it hits the cache: ` if ('$cachedRequired' in _this) {`. If I disable this branch the repro code from first post no longer fails.

The issue is simple: this cache is **shared** between model instances because it's stored in the schema. This is wrong because `required` function may refer to the document, so its return value cannot be cached between documents.

I'm not familiar with mongoose internals and I don't know why this caching was added in the first place, so I don't know what's the proper fix. I think it's a serious issue though and should be prioritized, because this bug is effectively breaking any `required` validator that's referring to `this`.

Cheers,
Sebastian
